tests/test_toolchain.py

```diff
@@ -0,0 +1,62 @@
+import sys
+import pytest
+import mock
+from pythonforandroid.toolchain import ToolchainCL
+from pythonforandroid.util import BuildInterruptingException
+
+
+class TestToolchainCL:
+
+    def test_help(self):
+        """
+        Calling with `--help` should print help and exit 0.
+        """
+        argv = ['toolchain.py', '--help', '--storage-dir=/tmp']
+        with mock.patch('sys.argv', argv), pytest.raises(SystemExit) as ex_info, \
+                mock.patch('argparse.ArgumentParser.print_help') as m_print_help:
+            ToolchainCL()
```

Comment:
Wouldn't something like this be a little easier to understand:

```
subprocess.check_output([
    sys.executable, "-c", "from pythonforandroid.toolchain import main(); main", "--help"
], cwd=os.path.join(os.path.dirname(__file__), ".."))
```

Or is that only me? Mock is fine but if it's not needed I would still rather use standard methods
---
There is also a lot of repetition of this block in particular:
```
        with mock.patch('sys.argv', argv), pytest.raises(SystemExit) as ex_info, \
                mock.patch('argparse.ArgumentParser.print_help') as m_print_help:
            ToolchainCL()
```
No matter if we end up using mock or not, maybe a helper function for that would be useful?

Reply:
The subprocess is overkill to me, it's really too much top level integration test.
Regarding the patching helper, yes I agree, I actually do this in my projects and at work, but since it's not common in p4a I didn't dare to introduce it.
I'll then

Reply:
> it's really too much top level integration test.

I'll let you decide but just as a note: that seems like an irrelevant point to me, because isn't this what the entire purpose is? Check if the overall tool responds fine to `--help`? It seems a little to me *with this particular argument* you're trying to isolate a component here with no benefit other than possibly complicating the test, and also removing it further from the actual usage later. But of course there are other completely valid reasons to stick with your current solution, e.g. if you find it more readable (which I don't but I can see others maybe would)

But keep in mind my proposal would also possibly greatly reduce the maintenance burden if `toolchain.py` is ever heavily refactored to also adjust the test here, because it nicely stays out of how the functionality needs to be performed as long as it works right without a crash, which I think is an additional benefit!

But that's just my personal opinion of course and I don't think this should hold off a merge

Reply:
Yes I definitely don't share this point of view regarding readability, I find that `subprocess.check_output` absolutely ugly. Not to say about parsing eventual errors that it could output or the fact that's unclear which python executable it would run, the fact that you need to play around with the cwd. For me it's just a big nope. I've hardly ever seen people writing unit tests calling `subprocess` :confused:

Reply:
it's `sys.executable` do people not usually know that? :smile: but it's fine, leave it with mocking I am clearly making you confused :joy: (I'm weird that's normal)